ctru-rs/src/applets/swkbd.rs

```diff
@@ -311,6 +383,80 @@ impl SoftwareKeyboard {
         self.state.filter_flags = filters.bits();
     }
 
+    /// Configure a custom filtering function to validate the input.
+    ///
+    /// The callback function must return a [`CallbackResult`] and the error message to display when the input is invalid.
+    ///
+    /// # Notes
+    ///
+    /// Passing [`None`] will unbind the custom filter callback.
+    ///
+    /// The error message returned by the callback should be shorter than `256` characters, otherwise it will be truncated.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # let _runner = test_runner::GdbRunner::default();
+    /// # fn main() {
+    /// #
+    /// use std::borrow::Cow;
+    /// use std::ffi::CString;
+    /// use ctru::applets::swkbd::{SoftwareKeyboard, CallbackResult};
+    ///
+    /// let mut keyboard = SoftwareKeyboard::default();
+    ///
+    /// keyboard.set_filter_callback(Some(Box::new(|str| {
+    ///     if str.to_str().unwrap().contains("boo") {
+    ///         return (
+    ///             CallbackResult::Retry,
+    ///             Some(Cow::Owned(CString::new("Ah, you scared me!").unwrap())),
+    ///         );
+    ///     }
+    ///
+    ///     (CallbackResult::Ok, None)
+    /// })));
+    /// #
+    /// # }
+    pub fn set_filter_callback(&mut self, callback: Option<Box<CallbackFunction>>) {
+        self.callback = callback;
+    }
+
+    /// Internal function called by the filter callback.
+    extern "C" fn internal_callback(
+        user: *mut libc::c_void,
+        pp_message: *mut *const libc::c_char,
+        text: *const libc::c_char,
+        _text_size: libc::size_t,
+    ) -> ctru_sys::SwkbdCallbackResult {
+        let this: *mut SoftwareKeyboard = user.cast();
+
+        unsafe {
+            // Reset any leftover error message.
+            (*this).error_message = None;
+
+            let text = CStr::from_ptr(text);
+
+            let result = {
+                // Run the callback if still available.
+                if let Some(callback) = &mut (*this).callback {
+                    let (res, cstr) = callback(text);
+
+                    (*this).error_message = cstr;
```

Comment:
nit: Add comments to explain why we have to store the error message and then reference it in this way.